pkg/sqlutils/transactions.go

```diff
@@ -2,37 +2,79 @@
 package sqlutils
 
 import (
+	"context"
 	"fmt"
 
 	"github.com/jmoiron/sqlx"
-)
+	"go.uber.org/zap"
 
-// TransactionFunc defines the function signature needed to represent passing a transaction and returning a generic type
-type TransactionFunc[T any] func(*sqlx.Tx) (*T, error)
+	"github.com/cmsgov/easi-app/pkg/appcontext"
+)
 
-// WithTransaction is a wrapper function which handles creating, committing or rolling back a transaction
+// WithTransactionRet is a wrapper function which handles creating, committing or rolling back a transaction
 // If there are any errors when executing the txFunc, the tx is rolled back. Otherwise, the tx is committed.
-func WithTransaction[T any](txPrep TransactionPreparer, txFunc TransactionFunc[T]) (*T, error) {
+// We use named return parameters here so that the deferred function can access and reassign the variables
+func WithTransactionRet[T any](ctx context.Context, txPrep TransactionPreparer, txFunc func(*sqlx.Tx) (T, error)) (result T, txErr error) {
+	logger := appcontext.ZLogger(ctx)
+
 	tx, err := txPrep.Beginx()
 	if err != nil {
-		return nil, fmt.Errorf("error creating transaction %w", err)
+		return result, fmt.Errorf("error creating transaction %w", err)
 	}
 
-	result, errFunc := txFunc(tx)
-	if errFunc != nil {
-		rollbackErr := tx.Rollback()
-		if rollbackErr != nil {
-			return nil, fmt.Errorf("issue rolling back transaction: %w", rollbackErr)
+	defer func() {
+		var defaultT T
+		if p := recover(); p != nil {
+			// if `p` is not nil, it means we panicked at some point in the `txFunc`
+			// attempt rollback
+			if err := tx.Rollback(); err != nil {
+				logger.Error("failed to rollback transaction after panic", zap.Error(err))
+			}
+
+			// empty result
+			result = defaultT
```

Comment:
@StevenWadeOddball @ClayBenson94 @mynar7 please see updates here (and line 49, 59)

we zero `result` in the event of an error so that the transaction will never return unusable data as `result` along with an error

Great point by @StevenWadeOddball - thank you for pointing that out!